internal/gke/update.go

```diff
@@ -0,0 +1,439 @@
+				&gkeapi.CidrBlock{
+					CidrBlock:   v.CidrBlock,
+					DisplayName: v.DisplayName,
+				})
+		}
+		needsUpdate = true
+	}
+	if needsUpdate {
+		logrus.Infof("updating master authorized networks configuration for cluster [%s]", config.Name)
+		_, err := client.Projects.
+			Locations.
+			Clusters.
+			Update(
+				ClusterRRN(config.Spec.ProjectID, config.Spec.Region, config.Spec.ClusterName),
+				&gkeapi.UpdateClusterRequest{
+					Update: clusterUpdate,
+				},
+			).Context(ctx).
+			Do()
+		if err != nil {
+			return NotChanged, err
+		}
+		return Changed, nil
+	}
+	return NotChanged, nil
+}
+
+// UpdateLoggingMonitoringService updates both LoggingService and MonitoringService.
+// In most cases, updating one requires explicitly updating the other as well, so these are paired.
+func UpdateLoggingMonitoringService(
+	ctx context.Context,
+	client *gkeapi.Service,
+	config *gkev1.GKEClusterConfig,
+	upstreamSpec *gkev1.GKEClusterConfigSpec) (Status, error) {
+	clusterUpdate := &gkeapi.ClusterUpdate{}
+	needsUpdate := false
+	if config.Spec.LoggingService != nil {
+		loggingService := *config.Spec.LoggingService
+		if loggingService == "" {
+			loggingService = CloudLoggingService
+		}
+		if *upstreamSpec.LoggingService != loggingService {
+			clusterUpdate.DesiredLoggingService = loggingService
+			needsUpdate = true
+		}
+	}
+	if config.Spec.MonitoringService != nil {
+		monitoringService := *config.Spec.MonitoringService
+		if monitoringService == "" {
+			monitoringService = CloudMonitoringService
+		}
+		if *upstreamSpec.MonitoringService != monitoringService {
+			clusterUpdate.DesiredMonitoringService = monitoringService
+			needsUpdate = true
+		}
+	}
+	if needsUpdate {
+		logrus.Infof("updating logging and monitoring configuration for cluster [%s]", config.Name)
+		_, err := client.Projects.
+			Locations.
+			Clusters.
+			Update(
+				ClusterRRN(config.Spec.ProjectID, config.Spec.Region, config.Spec.ClusterName),
+				&gkeapi.UpdateClusterRequest{
+					Update: clusterUpdate,
+				},
+			).Context(ctx).
+			Do()
+		if err != nil {
+			return NotChanged, err
+		}
+		return Changed, nil
+	}
+	return NotChanged, nil
+}
+
+// UpdateNetworkPolicyEnabled updates the Cluster NetworkPolicy setting.
+func UpdateNetworkPolicyEnabled(
+	ctx context.Context,
+	client *gkeapi.Service,
+	config *gkev1.GKEClusterConfig,
+	upstreamSpec *gkev1.GKEClusterConfigSpec) (Status, error) {
+	if config.Spec.NetworkPolicyEnabled == nil {
+		return NotChanged, nil
+	}
+
+	if *upstreamSpec.NetworkPolicyEnabled != *config.Spec.NetworkPolicyEnabled {
+		logrus.Infof("updating network policy for cluster [%s]", config.Name)
+		_, err := client.Projects.
+			Locations.
+			Clusters.
+			SetNetworkPolicy(
+				ClusterRRN(config.Spec.ProjectID, config.Spec.Region, config.Spec.ClusterName),
+				&gkeapi.SetNetworkPolicyRequest{
+					NetworkPolicy: &gkeapi.NetworkPolicy{
+						Enabled:  *config.Spec.NetworkPolicyEnabled,
+						Provider: NetworkProviderCalico,
```

Comment:
We always hardcode to Calico ?

Reply:
GKE only supports Calico. I figure if they add support for other CNIs later then we can rework this, but for the moment there's only one valid configuration for NetworkPolicy so it's easier to hardcode it than to validate it.